packages/components/src/List/Toolbar/Filter/Filter.component.js

```diff
@@ -1,68 +1,161 @@
 import React, { PropTypes } from 'react';
+import classNames from 'classnames';
 import DebounceInput from 'react-debounce-input';
 import FormControl from 'react-bootstrap/lib/FormControl';
-import Button from 'react-bootstrap/lib/Button';
+import { Action } from '../../../Actions';
 import Icon from '../../../Icon';
+import theme from './Filter.scss';
+
+const ESC_KEY = 27;
+const ENTER_KEY = 13;
+
+function onKeyDown(event, escAction, enterAction) {
+	switch (event.keyCode) {
+	case ESC_KEY:
+		escAction(event);
+		break;
+	case ENTER_KEY:
+		if (enterAction) {
+			enterAction(event);
+		}
+		break;
+	default:
+		break;
+	}
+}
+function FilterInput(props) {
+	const {
+		id,
+		debounceMinLength,
+		debounceTimeout,
+		onBlur,
+		onFocus,
+		onFilter,
+		onToggle,
+	} = props;
+
+	const inputProps = {
+		id,
+		name: 'search',
+		type: 'search',
+		placeholder: 'Filter',
+		className: theme.search,
+		'aria-label': 'Filter',
+		onBlur: onBlur && (event => onBlur(event, event.target.value)),
+		onFocus: onFocus && (event => onFocus(event, event.target.value)),
+		onChange: event => onFilter(event, event.target.value),
+		onKeyDown: event => onKeyDown(event, onToggle, onBlur),
+		autoFocus: true,
+	};
+
+	if (debounceMinLength || debounceTimeout) {
+		return (<DebounceInput
+			{...inputProps}
+			element={FormControl}
+			minLength={debounceMinLength}
+			debounceTimeout={debounceTimeout}
+		/>);
+	}
+
+	return <FormControl {...inputProps} />;
+}
+FilterInput.propTypes = {
+	id: PropTypes.string,
+	debounceMinLength: PropTypes.number,
+	debounceTimeout: PropTypes.number,
+	onBlur: PropTypes.func,
+	onFocus: PropTypes.func,
+	onFilter: PropTypes.func.isRequired,
+	onToggle: PropTypes.func,
+};
 
 /**
  * @param {object} props react props
  * @example
- <Filter name="Hello world"></Filter>
+ <Filter id="my-filter" docked="false" onFilter="filter()"></Filter>
  */
-function Filter({ id, onFilter, debounceMinLength, debounceTimeout }) {
-	function onFilterHandler(event) {
-		return onFilter(event.target.value, event);
+function Filter(props) {
+	const {
+		id,
+		debounceMinLength,
+		debounceTimeout,
+		docked,
+		highlight,
+		onBlur,
+		onFocus,
+		onFilter,
+		onToggle,
+	} = props;
+	if (docked) {
+		return (
+			<Action
+				id={id}
+				className="navbar-right"
+				onClick={onToggle}
+				label="Toggle filter"
+				hideLabel
+				icon="talend-search"
+				bsStyle="link"
+			/>
+		);
 	}
 
 	function onSubmit(event) {
 		event.preventDefault();
-		return onFilter(this.value, event);
+		return onFilter(event, event.target.search.value);
```

Comment:
I think using `_.get` would be a good thing here. You can't be sure the entire `event.target.search.value` exists.